### PR TITLE
feat: implement `PayloadProof` function

### PR DESCRIPTION
**Pull request by rupam-04**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implement a function to calculate the proof of the `ExecutionPayload` field in `BeaconBlockBody` (needed for light client support)

### CODE REVIEW

File: consensus-types/blocks/proofs.go [hunk 1]
```diff
@@ -172,3 +172,43 @@ func ComputeBlockBodyFieldRoots(ctx context.Context, blockBody *BeaconBlockBody)
 
 	return fieldRoots, nil
 }
+
+func ComputeBeaconBlockFieldRoots(ctx context.Context, block *BeaconBlock) ([][]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "ComputeFieldRootsWithHasher")
```
Comment (rkapka) [inline consensus-types/blocks/proofs.go:177]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "blocks.ComputeBlockFieldRoots")
```

File: consensus-types/blocks/proofs.go [hunk 2]
```diff
@@ -172,3 +172,43 @@ func ComputeBlockBodyFieldRoots(ctx context.Context, blockBody *BeaconBlockBody)
 
 	return fieldRoots, nil
 }
+
+func ComputeBeaconBlockFieldRoots(ctx context.Context, block *BeaconBlock) ([][]byte, error) {
```
Comment (rkapka) [inline consensus-types/blocks/proofs.go:176]: Can you rename it to `ComputeBlockFieldRoots` (without the `Beacon` part) just to be consistent with the other function?

File: consensus-types/blocks/proofs.go [hunk 3]
```diff
@@ -172,3 +172,43 @@ func ComputeBlockBodyFieldRoots(ctx context.Context, blockBody *BeaconBlockBody)
...
+	ctx, span := trace.StartSpan(ctx, "ComputeFieldRootsWithHasher")
+	defer span.End()
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
```
Comment (rkapka) [inline consensus-types/blocks/proofs.go:181]:
```suggestion
```
As discussed while implementing the previous function, this is not needed

File: beacon-chain/state/fieldtrie/field_trie_helpers.go [hunk 4]
```diff
@@ -328,3 +330,31 @@ func handleBalanceMVSlice(mv multi_value_slice.MultiValueSliceComposite[uint64],
 	}
 	return [][32]byte{}, nil
 }
+
+func PayloadProof(ctx context.Context, block *blocks.BeaconBlock) ([][]byte, error) {
```
Comment (rkapka) [inline beacon-chain/state/fieldtrie/field_trie_helpers.go:334]: - Let's put this in the same file as the other two. This is the state package.
- This function only makes sense when `block.version >= version.Bellatrix`. Phase0 and Altair forks didn't have a payload. We should check this at the beginning.

File: beacon-chain/state/fieldtrie/field_trie_helpers.go [hunk 5]
```diff
@@ -328,3 +330,31 @@ func handleBalanceMVSlice(mv multi_value_slice.MultiValueSliceComposite[uint64],
...
+	i := block.Body()
+	blockBody, ok := i.(*blocks.BeaconBlockBody)
```
Comment (rkapka) [inline beacon-chain/state/fieldtrie/field_trie_helpers.go:336]: If you put this in the `blocks` package, you should be able to write `block.Body` without casting, I think

File: beacon-chain/state/fieldtrie/field_trie_helpers.go [hunk 6]
```diff
@@ -328,3 +330,31 @@ func handleBalanceMVSlice(mv multi_value_slice.MultiValueSliceComposite[uint64],
...
+		return nil, errors.New("failed to cast block body")
+	}
+
+	blockBodyFieldRoots, err := blocks.ComputeBlockBodyFieldRoots(ctx, blockBody)
+	if err != nil {
+		return nil, err
+	}
+
+	blockBodyFieldRootsTrie := stateutil.Merkleize(blockBodyFieldRoots)
+	blockBodyProof := ProofFromMerkleLayers(blockBodyFieldRootsTrie, 9)
```
Comment (rkapka) [inline beacon-chain/state/fieldtrie/field_trie_helpers.go:347]: Can you create a `payloadFieldIndex` constant?

File: beacon-chain/state/fieldtrie/field_trie_helpers.go [hunk 7]
```diff
@@ -328,3 +330,31 @@ func handleBalanceMVSlice(mv multi_value_slice.MultiValueSliceComposite[uint64],
...
+
+	beaconBlockFieldRoots, err := blocks.ComputeBeaconBlockFieldRoots(ctx, block)
+	if err != nil {
+		return nil, err
+	}
+
+	beaconBlockFieldRootsTrie := stateutil.Merkleize(beaconBlockFieldRoots)
+	beaconBlockProof := ProofFromMerkleLayers(beaconBlockFieldRootsTrie, 4)
```
Comment (rkapka) [inline beacon-chain/state/fieldtrie/field_trie_helpers.go:355]: Can you create a `bodyFieldIndex` constant?

File: consensus-types/blocks/proofs.go (hunk 1 above)
Comment (rupam-04) [reply consensus-types/blocks/proofs.go:177]: > 

ah my bad

File: beacon-chain/state/fieldtrie/field_trie_helpers.go (hunk 4 above)
Comment (rupam-04) [reply beacon-chain/state/fieldtrie/field_trie_helpers.go:334]: > * Let's put this in the same file as the other two. This is the state package.
> 
>     * This function only makes sense when `block.version >= version.Bellatrix`. Phase0 and Altair forks didn't have a payload. We should check this at the beginning.

putting this in the blocks package requires me to import `fieldtrie` which causes an import cycle error

File: container/trie/sparse_merkle.go [hunk 8]
```diff
@@ -259,3 +259,18 @@ func (m *SparseMerkleTrie) NumOfItems() int {
 	}
 	return len(m.originalItems)
 }
+
+// ProofFromMerkleLayers creates a proof starting at the leaf index of the state Merkle layers.
```
Comment (rkapka) [inline container/trie/sparse_merkle.go:263]:
```suggestion
// ProofFromMerkleLayers creates a proof starting at the leaf index of the merkle layers.
```

**Review by rkapka — APPROVED** [review]

**Pull request merged**